### ISSUE TITLE
RapidJSON causes cc1plus: internal compiler error: Segmentation fault

### ISSUE THREAD

**Issue by Dono-Harjanto** [Tencent/rapidjson#860]
When cross-compiling the following code, the compiler crashes:
_In member function 'typename rapidjson::GenericMemberIterator<false, Encoding, Allocator>::Iterator rapidjson::GenericValue<Encoding, Allocator>::FindMember(const rapidjson::GenericValue<Encoding, SourceAllocator>&) [with SourceAllocator = rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator>, Encoding = rapidjson::UTF8<char>, Allocator = rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator>]':
cc1plus: internal compiler error: Segmentation fault_

The compile option:
arm-uclinuxeabi-g++ -mcpu=cortex-m3 -mthumb -ffunction-sections -fdata-sections   -c -O2

The code:
    Document doc;
    if (doc.Parse<0>(json).HasParseError())
    {
        // This is somehow not a valid JSON, bails out
        return -1;
    }
    std::map<std::string, std::string> m_deviceMetaMap;
    if (doc.HasMember("meta"))
    {
        const Value& deviceMeta = doc["meta"];
        if (deviceMeta.IsArray())
        {
            //for (Value::ConstValueIterator itr = deviceMeta.Begin(); itr != deviceMeta.End(); ++itr)
            for (SizeType i = 0; i < deviceMeta.Size(); i++)
            {
                /*
                const Value& name = (*itr)["name"];
                const Value& value = (*itr)["value"];
                std::string nameStr = name.GetString();
                std::string valueStr = value.GetString();
                 */
                const Value& nvp = deviceMeta[i];
                std::string nameStr = nvp["name"].GetString();
                std::string valueStr = nvp["value"].GetString();
                m_deviceMetaMap.insert(std::make_pair(nameStr, valueStr));
            }
        }
    }

Is this a bug in RapidJSON or a bug in compiler? Other compiler works fine, FYI.

**Comment by miloyip** [comment]
Probably compiler bug. You can try changing -O2 to -O1 or -O0 as workaround

**Comment by Dono-Harjanto** [comment]
Thank you for your reply. I did try changing optimization to O1, but no luck. However, changing optimization to O0 got it to compile with no crash.

One thing worth mentioning is that the old rapidjson code (downloaded back in 2015) compiled just fine even with O2 optimization. I attached the old rapidjson code here just as a reference.
[rapidjson-master.zip](https://github.com/miloyip/rapidjson/files/790905/rapidjson-master.zip)
